### PR TITLE
[FIX] website: Ensure no public users when archiving company

### PR DESCRIPTION
**Pull request by SergeBayet**
This commit prohibits the archival of a company if it is associated with
a website. Otherwise, the website wouldn't be accessible anymore by
public users (and so by all users if they need to login).

Step to reproduce:
- On website 2, change the company from SF to Chicago
- In the website list view (in debug mode), reorder the websites so
  website 2 is the first one in the tree view
- Archive the Chicago company (from debug > companies and do it from the
  list view)
- Try to access the website/runbot from an incognito tab
- It will show a raw 403 error

Note that the companies can only be archives since Odoo 16, thanks to
this commit: https://github.com/odoo/odoo/commit/7d0996bb151f68647b22719eee5ed4a4c35574bb

opw-3749772

### CODE REVIEW

File: addons/website/models/res_company.py [hunk 1]
```diff
@@ -19,6 +20,17 @@ def action_open_website_theme_selector(self):
...
+    @api.constrains('active')
+    def _check_active(self):
+        super()._check_active()
+        for company in self:
+            if not company.active:
+                public_users_count = self._get_public_users_count_for_website()
+                if public_users_count:
+                    raise ValidationError(_(
+                        'The company %(company_name)s cannot be archived because it has some public users', company_name=company.name
+                    ))
```
Comment (rdeodoo) [inline addons/website/models/res_company.py:32]: It would be about right (except that it's not I think, see below I'll explain why), but it should also consider the case of having a website_id set on the company but no public user linked to this company (regardless if this is a legit flow or not, it wouldn't cost much?)

That part would just be about
```py
if company.website_id:
    # this company is linked to a website
```

And it's actually all it should check, as it will implicitly also ensure that the user doesn't keep a website `user_id` linked to that company. Indeed:
- When creating a website or modifying it, the code ensure the fact that a public user linked to the company of that website is set.
- So, if you change the company of a website, it will also change the user_id (public user field) of that website
- Then, if you prevent the archive of company if the company is linked to a website, it would mean that the user will have to change the website company, and doing so it will also change the public user.

At the end, unless I am missing something, just checking for `company.website_id:` will cover everything.

File: addons/website/models/res_company.py [hunk 2]
```diff
@@ -27,6 +39,14 @@ def google_map_link(self, zoom=8):
         partner = self.sudo().partner_id
         return partner and partner.google_map_link(zoom) or None
 
+    def _get_public_users_count_for_website(self):
+        self.ensure_one()
+        # We need sudo to be able to see public users from others companies too
+        public_users = self.env.ref('base.group_public').sudo().with_context(active_test=False).users
+        public_users_for_website = public_users.filtered(lambda user: user.company_id == self)
+
+        return len(public_users_for_website)
```
Comment (rdeodoo) [inline addons/website/models/res_company.py:48]: Probably code to remove (see other comment), but to let you know: instead of copy pasting those 2 lines (containing sudo & comment), I'd extract those 2 lines from `_get_public_user` it in a new method and call it from your new `_check_active` and from the existing `_get_public_user`

**Review by rdeodoo — CHANGES_REQUESTED** [review]
Could you please also add a python unit test for this case? Maybe `test_res_users.py` is a good fit to write the test, but feel free to check if there is a better place.

File: addons/website/models/res_company.py (hunk 1 above)
Comment (SergeBayet) [reply addons/website/models/res_company.py]: @rdeodoo  Ok thanks for the explanation. I've simplified the commit.

**Comment by SergeBayet** [comment]
@rdeodoo A test has been added.

**Comment by rdeodoo** [comment]
Nice, quick, any reason to not target 15?

**Comment by SergeBayet** [comment]
@rdeodoo The bug-fix opw mentionned 16.x .. I didn't pay attention. Do I rebase on 15.0?

File: addons/website/tests/test_res_users.py [hunk 3]
```diff
@@ -102,3 +102,11 @@ def test_multi_website_multi_company(self):
 
         self._create_and_check_portal_user(False, company_1, company_2, website_1, website_2)
         self._create_and_check_portal_user(True, company_1, company_2, website_1, website_2)
+
+    def test_archive_company_linked_to_website(self):
+        company = self.env['res.company'].create({'name': 'Company'})
+        self.env['website'].create({'name': "Website 1", 'company_id': company.id})
+
+        # The company cannot be archived because it has a website linked to it
+        with self.assertRaisesRegex(ValidationError, 'The company Company cannot be archived'):
```
Comment (rdeodoo) [inline addons/website/tests/test_res_users.py:111]:
```suggestion
        with self.assertRaises(ValidationError, 'The company Company cannot be archived'):
```
What's the difference? It's actually the first time I see `assertRaisesRegex`, is there any reason to use it instead of `assertRaises`?

File: addons/website/models/res_company.py [hunk 4]
```diff
@@ -19,6 +20,15 @@ def action_open_website_theme_selector(self):
         action['target'] = 'new'
         return action
 
+    @api.constrains('active')
+    def _check_active(self):
+        super()._check_active()
+        for company in self:
+            if not company.active and company.website_id:
+                raise ValidationError(_(
+                    'The company %(company_name)s cannot be archived because it has a linked website', company_name=company.name
```
Comment (rdeodoo) [inline addons/website/models/res_company.py:29]: It needs to be added inside the `.pot` file.
Maybe worth mentioning which website in the error message, and maybe worth telling "Change that website's company first", no?

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/website/tests/test_res_users.py (hunk 3 above)
Comment (SergeBayet) [reply addons/website/tests/test_res_users.py:111]: I must admit that since it was my first time creating a Python test, I drew significant inspiration from `odoo/addons/base/tests/test_res_company.py`
```# The user cannot be unarchived because its default company is archived
with self.assertRaisesRegex(ValidationError, 'Company foo is not in the allowed companies'):
       user.action_unarchive()
 ```
 But indeed you're right, it should be easier with assertRaises.

Comment (SergeBayet) [reply addons/website/tests/test_res_users.py:111]: @rdeodoo  With further investigation, it seems to be better with assertRaisesRegex so we can check the content of the error text field, and not only the Error Type, that can be another ValidationError. Tell me if I'm wrong, but browsing the code base, I didn't see a self.assertRaises that fits my case.

File: addons/website/models/res_company.py (hunk 4 above)
Comment (SergeBayet) [reply addons/website/models/res_company.py:29]: Do I have to add it manually in the .pot file? It seems to be automated...

Comment (rdeodoo) [reply addons/website/models/res_company.py]: It's not automated, `.pot` are never touched anymore (except from our manual diff) once a branch has been released.

See https://discord.com/channels/803943906625716244/981917722826903572/1161603743599579207 and https://github.com/odoo/odoo/pull/113570#pullrequestreview-1313445887
It should give you a clear explanation of the `.po`, `.pot` and transifex processes.

Don't hesitate to ask if it needs more clarification

File: addons/website/i18n/website.pot [hunk 5]
```diff
@@ -9203,6 +9191,15 @@ msgstr ""
...
 msgstr ""
 
+#. module: website
+#. odoo-python
+#: code:addons/website/models/res_company.py:0
+#, python-format
+msgid ""
+"The company %(company_name)s cannot be archived because it has a linked website %(website_name)s \n"
+"Change that website's company first."
+msgstr ""
```
Comment (rdeodoo) [inline addons/website/i18n/website.pot:9201]: When adding stuff inside `.pot`, you need to limit the diff to strictly what you added. Could you remove all other diff in this file?


If you removed something (eg you replaced a term, which should almost never happen, so in the `pot` it's actually a removal and an addition), you never remove it from the `.pot`, the next master release will re-sync everything.

It's important because if you remove something from the `.pot` because the codebase doesn't need it anymore, some DBs will still use the translation, like if they COW'd a view or something.
Probably also for custos etc

tl;dr: just add the term you added in the codebase, exactly where it should be if you export it (so like here, in alphabetical order)

File: addons/website/i18n/website.pot [hunk 6]
```diff
@@ -6,8 +6,8 @@ msgid ""
 msgstr ""
 "Project-Id-Version: Odoo Server 16.0\n"
 "Report-Msgid-Bugs-To: \n"
-"POT-Creation-Date: 2024-02-06 13:31+0000\n"
-"PO-Revision-Date: 2024-02-06 13:31+0000\n"
+"POT-Creation-Date: 2024-03-06 13:53+0000\n"
+"PO-Revision-Date: 2024-03-06 13:53+0000\n"
```
Comment (rdeodoo) [inline addons/website/i18n/website.pot:10]: only exception could be the revision-date, you can keep it on top of your real diff if you want, it's not mandatory it has no real purpose, and it will make your forward port harder but as you prefer for this one

File: addons/website/models/res_company.py [hunk 7]
```diff
@@ -19,6 +20,18 @@ def action_open_website_theme_selector(self):
         action['target'] = 'new'
         return action
 
+    @api.constrains('active')
+    def _check_active(self):
+        super()._check_active()
+        for company in self:
+            if not company.active and company.website_id:
+                raise ValidationError(_(
+                    'The company %(company_name)s cannot be archived because it has a linked website %(website_name)s '
```
Comment (rdeodoo) [inline addons/website/models/res_company.py:29]:
```suggestion
                    'The company %(company_name)s cannot be archived because it has a linked website %(website_name)s.'
```
No?

**Review by rdeodoo — CHANGES_REQUESTED** [review]

File: addons/website/tests/test_res_users.py (hunk 3 above)
Comment (rdeodoo) [reply addons/website/tests/test_res_users.py]: I don't think it's worth ensuring the message is the correct one, if we ensure the operation doesn't go through, it seems more than enough to me.
Especially since it means copying the exception message in the test, having to keep them sync'd forever.

But why not.. Feel free to use what you prefer here, personally I'd just go with 
```py
with self.assertRaises(ValidationError):
```

**Comment by SergeBayet** [comment]
@rdeodoo Ready for review.

File: addons/website/models/res_company.py [hunk 8]
```diff
@@ -19,6 +20,18 @@ def action_open_website_theme_selector(self):
         action['target'] = 'new'
         return action
 
+    @api.constrains('active')
+    def _check_active(self):
+        super()._check_active()
+        for company in self:
+            if not company.active and company.website_id:
+                raise ValidationError(_(
+                    'The company %(company_name)r cannot be archived because it has a linked website %(website_name)r.'
```
Comment (rdeodoo) [inline addons/website/models/res_company.py:29]: (I already changed it so I can't comment the old code, but here is the diff)
```diff
-                    'The company %(company_name)s cannot be archived because it has a linked website %(website_name)s.'
+                    'The company %(company_name)r cannot be archived because it has a linked website %(website_name)r.'
```
Generally a quick win regarding UX/UI:
```
The company Company cannot be archived because it has a linked website Website 1.
=>
The company 'Company' cannot be archived because it has a linked website 'Website 1'.
```

**Review by rdeodoo — APPROVED** [review]
@robodoo r+ thanks for the changes :+1: 

I push forced to adapt the commit message and:
- (really mandatory) explain why it's a problem, you just wrote `This commit prohibits the archival of a company if it is associated with a website.` but it was not explaining that it was actually really critical since the DB would not be accessible anymore after that for public users (and so any users actually if not logged in yet).
- (mandatory) include the steps to reproduce the bug
- (nice to have) include https://github.com/odoo/odoo/commit/7d0996bb151f68647b22719eee5ed4a4c35574bb in your commit message

I also replace `%s` by `%r` which is important here, otherwise it can be hard to understand the error message regarding your company and/or site name which is included in the sentence without clearly be marked as such

The more info in the commit message the better :+1: 

image
